backend/main/chapters/c07_nested_loops.py

```diff
@@ -399,14 +399,62 @@ def solution(self, players: List[str]):
             """
              )]
 
-    class crack_password_exercise(ExerciseStep):
+    class player_vs_player_bonus(ExerciseStep):
         """
 Well done! If you'd like a bonus challenge, remove the repetition and only output each pair of players once, e.g:
 
     Alice vs Bob
     Alice vs Charlie
     Bob vs Charlie
 
+        """
+
+        def solution(self, players: List[str]):
+            for i in range(len(players)):
+                for j in range(i + 1, len(players)):
+                    print(players[i], 'vs', players[j])
+
+        hints = """
+This is similar to the previous exercise, you'll need a for loop inside a for loop like before.
+This time, in your loops you need to use `range` and `len`, and use indexing to access the list entries.
+Look at the desired output: `Alice vs Bob`: `Alice` comes before `Bob` in the `players` list.
+Similarly in `Alice vs Charlie` `Alice` comes before `Charlie`, and in `Bob vs Charlie` `Bob` comes before `Charlie` in `players`.
+We don't want to print `Bob vs Alice` because `Bob` comes AFTER `Alice` in `players`.
+The only pairs we want to print are those where the left player comes before the right player in the list.
```

Comment:
Ah, this is actually one of the disadvantages of a concrete exercise that needs to check that the output is exactly right (as opposed to letting the user decide for themselves in a soft bonus challenge). There's only one correct output. I hadn't thought about it here. The user might feel that `Bob vs Alice` is acceptable instead of `Alice vs Bob`. We need to specify the correct output precisely and clearly in the text, it can't only be hidden in hints. Ideally we want users to try and solve problems without looking at the hints, and they should always know exactly what the goal is.

Reply:
I thought about that, but wasn't sure how to go about it. Wanted to hear your take. Another potential issue is that since the names in the list are already in alphabetical order (but they don't have to be) the user might think that they only need to print pairs in alphabetical order (rather than list order). Maybe ruin the alphabetical order in the list, and ask the question that way? (then specifying the only pairs to be printed are the ones left-to-right on the list)

Reply:
Yes, ruin the alphabetical order. Good idea.

Reply:
Will do!